### ISSUE TITLE
Remove need for pub in url

### ISSUE THREAD

**Issue by randyzwitch** [tlienart/Franklin.jl#335]
I've been exploring JuDoc for a bit, and one thing I noticed that doesn't seem ideal is having the `pub` directory in the URL name. Jekyll manages this by having a folder `_site`, but not sure how it is re-writing things so that directory isn't part of the URL

**Comment by tlienart** [comment]
That's a fair request but I'm currently not sure how to  support it (help is welcome). I know it can be done with  `.htaccess` but that would only  be an option for people who are  using their own hosting.

If someone has a clear idea of how this works with Hugo or Jekyll, I'll be happy to  learn and try to implement something similar.

PS: what I like with the current state of things is that the structure is consistent locally, online, and on the repo. But I would be in favour of supporting an alternative way that looks like hugo/jekyll.

**Comment by randyzwitch** [comment]
Yeah, the more I look at it, it might be just a convention for GitHub Pages. Jekyll does this, as well as the R one (knitr maybe?). Of course, I recognize that not everyone is using GitHub Pages :)

I'll take a look and see if I can figure something out. I'd love to move from Jekyll to JuDoc, would make life a lot easier since I don't know ruby

**Comment by tlienart** [comment]
I think this: https://gist.github.com/cobyism/4730490 shows a step by  step  way of  doing things. I'll experiment with it (I want to make sure I get this right before rushing anything 😄).

**Comment by randyzwitch** [comment]
It appears this all works by having a subfolder with the "page name", then a file as `index.html`. Leaving the `index.html` off the end makes it have the nice like permalink

![image](https://user-images.githubusercontent.com/2762787/71750320-097d0c80-2e46-11ea-840b-e6de9ae72ddb.png)

In my git repo, these folders don't exist, so I suspect that GitHub Pages creates them behind the scenes with Jekyll. In implementing this ourselves, we'd have to check in the `folder/index.html` files, since GitHub wouldn't know how to build them.

**Comment by tlienart** [comment]
Thanks! so I think there's two things here:

* machinery to push stuff on `gh-pages` (that would solve the `/pub/` thing)
* machinery to have `/name/` instead of `/name.html` 

### first thing

Would require something like:
- generate all files in a `_site` subfolder with a proper structure
- have the `serve` function watch _that_ folder for changes
- push the content of  that `_site` folder to the `gh-pages` with the gymnastic suggested in one of the pages above
- deploy that `gh-pages`

### second thing

- have each page be an `index.html` of an appropriate folder
- use meta canonical tags

### Things to do

Remains for me to think about how this would work in practice so that I don't break everything for people who  are using things as they are now :)

**Comment by randyzwitch** [comment]
I’m still planning on taking a look. It seems like this could be part of the new site function, and then upon serve(), we could check which directory to serve from

**Comment by tlienart** [comment]
Awesome, thanks

**Comment by tlienart** [comment]
Ok I've now had the time to look into  this in some detail. It seems reasonably straightforward but will require significant changes to a few things. Below is  a full plan including how I think it can be made backward compatible so that people updating Franklin with an existing website can just keep using things as they were.

## New approach

The overview is as follows:

* have a `_site` folder where the generated content goes
* in that `_site` folder, every subpage has its own folder with an  `index.html`, this allows for URL like `username.github.io/root/page1` to be valid (no need to add `...html` at the end), `root` can either be empty (user website) or non-empty (project website)

Ex:

```
.
├── assets
├── css
├── index.html                    # landing page
├── menu1
│   └── index.html                # accessible at .../menu1/
```

* have `_site` to be on its own  `gh-pages` worktree following essentially: https://sangsoonam.github.io/2019/02/08/using-git-worktree-to-deploy-github-pages.html note that the name of the branch *has* to be `gh-pages` for GitHub
* with GitLab it's **different** in that stuff has to be in a dedicated `public/` folder; we could  use a script that pulls that from the gh-pages branch but that's maybe unnecessarily  complicated.
* with arbitrary provider (e.g. Netlify) again it's probably better to have a target folder in which everything is generated

So for non-Github deployment, there would need to be a global var to select an output dir, I'm yet to test this with netlify. 

### Changes to FranklinTemplates

* bring everything from `src/` one level up
* add a `_site` folder, put `assets/` and `libs/` in it
* general adjustments to  make the template deployment work with this

### Changes to Franklin

* read global var to see whether to use the `gh-pages` approach or a target folder approach
  * first approach: add `_site` as a worktree etc and add `_site` to the `.gitignore`
  * second approach, generate everything in `_site` and push
* adjust all paths specs to generate files in `_site` and pages in `/pagename/index.html` 

## Make stuff backward compatible

* add a `@def newdeploy = false` to `config.md`, everything stays the same as before.


Fingers crossed this will be done the coming week, I'll probably need patient folks to test it all works without breaking everything!

**Comment by tlienart** [comment]
This is now implemented in #356 modulo a few additional tests and some small utils to help  setup the `gh-pages` branch as necessary, I'm hopeful this will be  merged by the weekend.

**Comment by randyzwitch** [comment]
Wow, great work! I’m traveling now, but I hope to check this out soon. I’d love to convert my Jekyll blog, and if I can do that with Julia code, I’d be happy to make a converter plugin/package

**Comment by tlienart** [comment]
This is now closed by #356 and the 0.6 release. Feedback welcome!
